### PR TITLE
Make NNCFNetwork deepcopy-able

### PR DESCRIPTION
**Pull request by vshampor**
### Changes
Fixed a return value for the internal NNCFNetwork function that populates one of the fields, and made the threading-related fields of `TracingContext` only instantiate when the context is entered.

### Reason for changes

The NNCFNetwork object is not deepcopy-able otherwise.

### Related tickets

76848

### Tests

test_deepcopy_nncf_network

### CODE REVIEW

File: nncf/torch/dynamic_graph/graph_tracer.py [hunk 1]
```diff
@@ -98,8 +98,8 @@ def trace_graph(self, model: torch.nn.Module, context_to_use: Optional['TracingC
 
         context_to_use.enable_trace_dynamic_graph()
         from nncf.torch.utils import training_mode_switcher
-        context_to_use.base_module_thread_local_replica = model
         with context_to_use as _ctx:
+            _ctx.base_module_thread_local_replica = model
```
Comment (ljaljushkin) [inline nncf/torch/dynamic_graph/graph_tracer.py:102]: just for understanding, what was the main change fixing the deepcopy? I wonder whether all changes were required to fix it, or there is also some refactoring to handle thread_local replica more clear.

Comment (vshampor) [reply nncf/torch/dynamic_graph/graph_tracer.py:102]: My original fix seems to have been incorrect. I pushed another version, in which the main change fixing the deepcopy is separating the threading variables (which are non-deepcopyable) into a specific object with an overridden `__deepcopy__` method.

Comment (ljaljushkin) [reply nncf/torch/dynamic_graph/graph_tracer.py:102]: Does new test cover the case when previous implementation failed?

**Comment by ljaljushkin** [comment]
Jenkins please retry a build

File: nncf/torch/dynamic_graph/context.py [hunk 2]
```diff
@@ -319,18 +309,19 @@ def _init_thread_local(self):
         tl.operator_counters = {}
         tl.node_call_tracker = {}
 
+
```
Comment (alexsu52) [inline nncf/torch/dynamic_graph/context.py:312]:
```suggestion
```

File: tests/torch/test_nncf_network.py [hunk 3]
```diff
@@ -945,3 +945,10 @@ def get_ip_graph_for_test(nncf_graph: NNCFGraph,
                                    allowed_pre_hook_insertion_points=pre_hooks,
                                    allowed_post_hook_insertion_points=post_hooks)
     return ip_graph
+
```
Comment (alexsu52) [inline tests/torch/test_nncf_network.py:948]:
```suggestion


```

**Review by alexsu52 — APPROVED** [review]
Cool staff!

**Comment by vshampor** [comment]
Jenkins please retry a build

File: nncf/torch/dynamic_graph/graph_tracer.py (hunk 1 above)
Comment (vshampor) [reply nncf/torch/dynamic_graph/graph_tracer.py:102]: The failures of the previous implementation were caught by the rest of the tests.

**Review by ljaljushkin — APPROVED** [review]

**Pull request merged**